core/conn/unixodbc/odbc/odbcclient/unixcli/cli/ctosqlconv.cpp

```diff
@@ -1036,2500 +1166,3031 @@ unsigned long ODBC::ConvertCToSQL(SQLINTEGER	ODBCAppVersion,
+                        ulTmp = (ULONG_P)dTmp;
+                        if  (dTmp != ulTmp)
+                            retCode = IDS_01_S07;
+                        DataPtr = &ulTmp;
+                        DataLen = sizeof(ULONG_P);
+                    }
+                    else
+                    {
+                        if (!signedInteger)
+                        {
+                            if (dTmp < 0 || dTmp > UINT_MAX )
+                                return IDS_22_003;
+                            lTmp = (SLONG_P)dTmp;
+                        }
+                        else
+                        {
+                            if (dTmp < INT_MIN || dTmp > INT_MAX)
+                                return IDS_22_003;
+                            lTmp = (SLONG_P)dTmp;
+                            if  (dTmp != lTmp)
+                                retCode = IDS_01_S07;
+                        }
+                        DataPtr = &lTmp;
+                        DataLen = sizeof(SLONG_P);                    }
+                    break;
 
-				retCode = Ascii_To_Bignum_Helper(cTmpBuf,
-					strlen(cTmpBuf),
-					(char*)cTmpBuf2,
-					targetLength,
-					targetPrecision,
-					targetScale,
-					SQLDataType,
-					&dataTruncatedWarning);
+                case SQL_REAL:
+                    if (dTmp < -FLT_MAX || dTmp > FLT_MAX)
+                        return IDS_22_003;
+                    fltTmp = (SFLOAT)dTmp;
+                    DataPtr = &fltTmp;
+                    DataLen = sizeof(fltTmp);
+                    break;
 
-				if(retCode != SQL_SUCCESS)
-					return retCode;
+                case SQL_DOUBLE:
+                case SQL_FLOAT:
+                    DataPtr = &dTmp;
+                    DataLen = sizeof(dTmp);
+                    break;
 
-				useDouble = FALSE;
-				if (DataPtr == NULL)
-					DataPtr = cTmpBuf2;
+                case SQL_DECIMAL:
+                    if (targetPrecision >= sizeof(cTmpBuf))
+                        return IDS_22_003;
+#if defined MXHPUX || defined MXOSS || defined MXAIX || MXSUNSPARC || MXSUNSPARC
+                    dTmp1 = pow((double)10,targetPrecision-targetScale+1);
+#else
+                    dTmp1 = pow(10,targetPrecision-targetScale+1);
+#endif
+                    if (targetUnsigned)
+                    {
+                        if (dTmp < 0)
+                            return IDS_22_003_02;  //negValue in unsigned column
+                        if (dTmp > dTmp1)
+                            return IDS_22_003;
 
-				DataLen = targetLength;
-				memcpy(outDataPtr, DataPtr, DataLen);
-				if (byteSwap)
-				{
-					if (Datatype_Dependent_Swap((BYTE *)outDataPtr, SQLDataType, targetCharSet, DataLen, IEEE_TO_TANDEM) != STATUS_OK)
-						return IDS_HY_000;
-				}
+                        tempPtr = _fcvt(dTmp, targetScale, &dec, &sign);
+                        tempLen = strlen(tempPtr);
+                        tempLen1 = (short)(targetPrecision-tempLen);
```

Comment:
Not sure why the "(short)" cast is here. Could it lose information? (short)500000 = -24288, while (short)40000 is 6784.

Reply:
The value targetPrecision is small enough and in a range from 0-128 even for numeric

Reply:
OK.